src/engine/strat_engine/device.rs

```diff
@@ -0,0 +1,179 @@
+    /// If there is an error finding out the info, return that error.
+    /// Also, return an error if a device is not appropriate for this pool.
+    fn filter_devs<I>(dev_infos: I,
+                      pool_uuid: &PoolUuid,
+                      force: bool)
+                      -> EngineResult<Vec<(Device, (PathBuf, Bytes, File))>>
+        where I: Iterator<Item = (Device, EngineResult<(PathBuf, Bytes, DevOwnership, File)>)>
+    {
+        let mut add_devs = Vec::new();
+        for (dev, dev_result) in dev_infos {
+            let (devnode, dev_size, ownership, f) = try!(dev_result);
+            if dev_size < MIN_DEV_SIZE {
+                let error_message = format!("{} too small, minimum {} bytes",
+                                            devnode.display(),
+                                            MIN_DEV_SIZE);
+                return Err(EngineError::Engine(ErrorEnum::Invalid, error_message));
+            };
+            match ownership {
+                DevOwnership::Unowned => add_devs.push((dev, (devnode, dev_size, f))),
+                DevOwnership::Theirs => {
+                    if !force {
+                        let error_str = format!("First 8K of {} not zeroed", devnode.display());
+                        return Err(EngineError::Engine(ErrorEnum::Invalid, error_str));
+                    } else {
+                        add_devs.push((dev, (devnode, dev_size, f)))
+                    }
+                }
+                DevOwnership::Ours(uuid) => {
+                    if *pool_uuid != uuid {
+                        let error_str = format!("Device {} already belongs to Stratis pool {}",
+                                                devnode.display(),
+                                                uuid);
+                        return Err(EngineError::Engine(ErrorEnum::Invalid, error_str));
+                    } else {
+                        // Already in this pool (according to its header)
+                        // TODO: Check we already know about it
+                        // if yes, ignore. If no, add it w/o initializing?
+                    }
+                }
+            }
+        }
+        Ok(add_devs)
+    }
+
+    try!(validate_mda_size(mda_size));
+
+    let dev_infos = devices.into_iter().map(|d: Device| (d, dev_info(&d)));
+
+    let add_devs = try!(filter_devs(dev_infos, pool_uuid, force));
+
+    let mut bds = Vec::new();
+    for (dev, (devnode, dev_size, mut f)) in add_devs {
+
+        let bda = try!(BDA::initialize(&mut f,
+                                       pool_uuid,
+                                       &Uuid::new_v4(),
+                                       mda_size,
+                                       dev_size.sectors()));
+        let allocator = RangeAllocator::new_with_used(bda.dev_size(), &[(Sectors(0), bda.size())]);
+
+        bds.push(BlockDev::new(dev, devnode, bda, allocator));
+    }
+    Ok(bds)
```

Comment:
Last 12 lines in `initialize` here are extremely BlockDev-specific. Can this stuff stay in blockdev.rs, or something like that? Given the delineation we are now drawing between device.rs and blockdev.rs, having device.rs not using `BlockDev` at all seems attractive.